demos/http/http_demo_s3_download_async/http_demo_s3_download_async.c

```diff
@@ -0,0 +1,751 @@
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX includes. */
+#include <unistd.h>
+#include <fcntl.h>
+#include <mqueue.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <signal.h>
+
+/* Include Demo Config as the first non-system header. */
+#include "demo_config.h"
+
+/* Common HTTP demo utilities. */
+#include "http_demo_utils.h"
+
+/* HTTP API header. */
+#include "core_http_client.h"
+
+/* OpenSSL transport header. */
+#include "openssl_posix.h"
+
+/* Retry utilities. */
+#include "retry_utils.h"
+
+/* Check that TLS port of the server is defined. */
+#ifndef HTTPS_PORT
+    #error "Please define a HTTPS_PORT."
+#endif
+
+/* Check that a path for Root CA Certificate is defined. */
+#ifndef ROOT_CA_CERT_PATH
+    #error "Please define a ROOT_CA_CERT_PATH."
+#endif
+
+/* Check that a presigned url for the target S3 file is defined. */
+#ifndef S3_PRESIGNED_URL
+    #error "Please define a S3_PRESIGNED_URL."
+#endif
+
+/* Check that a path for HTTP Method GET is defined. */
+#ifndef S3_FILE_SIZE
+    #error "Please define a S3_FILE_SIZE."
+#endif
+
+/**
+ * @brief ALPN protocol name to be sent as part of the ClientHello message.
+ *
+ * @note When using ALPN, port 443 must be used to connect to AWS IoT Core.
+ */
+#define IOT_CORE_ALPN_PROTOCOL_NAME    "\x0ex-amzn-http-ca"
+
+/**
+ * @brief Delay in seconds between each iteration of the demo.
+ */
+#define DEMO_LOOP_DELAY_SECONDS        ( 5U )
+
+/* Check that transport timeout for transport send and receive is defined. */
+#ifndef TRANSPORT_SEND_RECV_TIMEOUT_MS
+    #define TRANSPORT_SEND_RECV_TIMEOUT_MS    ( 1000 )
+#endif
+
+/* Check that size of the user buffer is defined. */
+#ifndef USER_BUFFER_LENGTH
+    #define USER_BUFFER_LENGTH    ( 4096 )
+#endif
+
+/* Check that request queue name is defined. */
+#ifndef REQUEST_QUEUE
+    #define REQUEST_QUEUE    "/demo_request_queue"
+#endif
+
+/* Check that request queue name is defined. */
+#ifndef RESPONSE_QUEUE
+    #define RESPONSE_QUEUE    "/demo_response_queue"
+#endif
+
+/**
+ * @brief Field name of the HTTP Range header to read from server response.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD           "Content-Range"
+
+/**
+ * @brief Length of the HTTP Range header field.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH    ( sizeof( HTTP_CONTENT_RANGE_HEADER_FIELD ) - 1 )
+
+/* The location of the host address within string S3_PRESIGNED_URL. */
+static const char * pAddress = NULL;
+
+/* The host address string extracted from S3_PRESIGNED_URL. */
+static char serverHost[ MAX_HOST_ADRESS_LENGTH ];
+
+/* The length of the host address found in string S3_PRESIGNED_URL. */
+static size_t serverHostLength = 0;
+
+/**
+ * @brief Data type for request queue.
+ *
+ * In addition to sending the headers struct, we need to send the buffer it uses.
+ * The pointer to the buffer in the headers struct will be incorrect after
+ * it is received so will need to be fixed.
```

Comment:
```suggestion
 * In addition to sending #HTTPRequestHeaders_t, we need to send the buffer it uses.
 * The pointer to the buffer in #HTTPRequestHeaders_t will be incorrect after
 * it is received, so it will need to be fixed.
```